File: app/collector/cc/exe.go
package cc

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/ankeesler/btool/app/collector"
	"github.com/ankeesler/btool/log"
	"github.com/ankeesler/btool/node"
	"github.com/pkg/errors"
)

// Exe is a collector.Consumer that adds executable node.Node's.
type Exe struct {
	rf collector.ResolverFactory
}

// NewExe creates a new Exe.
func NewExe(rf collector.ResolverFactory) *Exe {
	return &Exe{
		rf: rf,
	}
}

// Consume will listen to executable node.Node's (i.e., node.Node's with no file
// extension) and create a link target for that node.Node. It walks the source
// tree backwards to find the dependent objects and libraries.
func (e *Exe) Consume(s collector.Store, n *node.Node) error {
	if filepath.Ext(n.Name) != "" {
		return nil
	}

	c := s.Get(n.Name + ".c")
	cc := s.Get(n.Name + ".cc")
	if c != nil && cc != nil {
		return fmt.Errorf("ambiguous exe %s: %s or %s?", n, c, cc)
	} else if c == nil && cc == nil {
		return fmt.Errorf("no known source for exe %s", n)
	}

	var err error
	objs := make([]*node.Node, 0)
	if c != nil {
		log.Debugf("exe %s has source %s", n, c)
		err = collectObjs(s, c, ".c", &objs)
	} else { // cc
		log.Debugf("exe %s has source %s", n, cc)
		err = collectObjs(s, cc, ".cc", &objs)
	}

	if err != nil {
		return errors.Wrap(err, "collect objs")
	}

	libraries := make([]*node.Node, 0)
	for _, obj := range objs {
		n.Dependency(obj)

		if err := collectLibraries(s, obj, &libraries); err != nil {
			return errors.Wrap(err, "collect libraries")
		}
	}
	for _, library := range libraries {
		n.Dependency(library)
	}

	var linkFlags []string
	linkFlags, err = CollectLinkFlags(n)
	if err != nil {
		return errors.Wrap(err, "collect link flags")
	}

	var r node.Resolver
	if c != nil {
		r = e.rf.NewLinkC(linkFlags)
	} else {
		r = e.rf.NewLinkCC(linkFlags)
	}
	n.Resolver = r

	s.Set(n)

	return nil
}

// The provided n should be a c/cc file!
func collectObjs(
	s collector.Store,
	n *node.Node,
	ext string,
	objs *[]*node.Node,
) error {
	obj := s.Get(strings.ReplaceAll(n.Name, ext, ".o"))
	if obj == nil {
		return fmt.Errorf("no know object for %s", n)
	}
	if contains(*objs, obj) {
		return nil
	}
	*objs = append(*objs, obj)
	log.Debugf("remembering object %s", obj)

	for _, d := range n.Dependencies {
		if !strings.HasSuffix(d.Name, ".h") {
			continue
		}

		src := s.Get(strings.ReplaceAll(d.Name, ".h", ext))
		if src == nil {
			log.Debugf("skipping header-only dependency %s", d.Name)
			continue
		}

		if err := collectObjs(s, src, ext, objs); err != nil {
			return err
		}
	}

	return nil
}

func collectLibraries(
	s collector.Store,
	n *node.Node,
	libraries *[]*node.Node,
) error {
	if err := node.Visit(n, func(vn *node.Node) error {
		var labels Labels
		if err := collector.FromLabels(vn, &labels); err != nil {
			return errors.Wrap(err, "from labels")
		}

		for _, library := range labels.Libraries {
			libraryN := s.Get(library)
			if libraryN == nil {
				return fmt.Errorf("unknown node for library %s", library)
			}

			*libraries = append(*libraries, libraryN)
		}

		return nil
	}); err != nil {
		return errors.Wrap(err, "visit")
	}

	return nil
}

// CollectLinkFlag will walk a node.Node graph and return all of the linker flags
// paths encoutered as a part of a node.Node's Labels along the way.
func CollectLinkFlags(n *node.Node) ([]string, error) {
	return collectLabels(n, func(l *Labels) []string { return l.LinkFlags })
}

func contains(nodes []*node.Node, n *node.Node) bool {
	for _, nn := range nodes {
		if nn == n {
			return true
		}
	}
	return false
}

File: cc/app/collector/cc/exe.cc
#include "app/collector/cc/exe.h"

#include <cassert>

#include <string>
#include <vector>

#include "app/collector/cc/properties.h"
#include "app/collector/properties.h"
#include "app/collector/store.h"
#include "log.h"
#include "node/node.h"
#include "util/fs/fs.h"
#include "util/string/string.h"
#include "util/util.h"

namespace btool::app::collector::cc {

void CollectObjects(::btool::app::collector::Store *s, ::btool::node::Node *n,
                    const std::string &ext,
                    std::vector<::btool::node::Node *> *objs);
void CollectLibraries(::btool::app::collector::Store *s, ::btool::node::Node *n,
                      std::vector<::btool::node::Node *> *libs);
void CollectLinkFlags(::btool::node::Node *n, std::vector<std::string> *flags);

void Exe::OnNotify(::btool::app::collector::Store *s, const std::string &name) {
  if (::btool::util::fs::Ext(name) != "") {
    return;
  }

  auto n = s->Get(name);
  if (n == nullptr) {
    return;
  }

  std::string ext = ".c";
  auto src = s->Get(name + ext);
  if (src == nullptr) {
    ext = ".cc";
    src = s->Get(name + ext);
    if (src == nullptr) {
      DEBUGS() << "cannot find source for exe " << name << std::endl;
      assert(0);
    }
  }

  std::vector<::btool::node::Node *> objs;
  CollectObjects(s, src, ext, &objs);
  for (auto obj : objs) {
    n->dependencies()->push_back(obj);
  }

  std::vector<::btool::node::Node *> libs;
  CollectLibraries(s, n, &libs);
  for (auto lib : libs) {
    n->dependencies()->push_back(lib);
  }

  std::vector<std::string> flags;
  CollectLinkFlags(n, &flags);

  auto r = (ext == ".cc" ? rf_->NewLinkCC(flags) : rf_->NewLinkC(flags));
  n->set_resolver(r);

  Notify(s, n->name());
}

void CollectObjects(::btool::app::collector::Store *s, ::btool::node::Node *n,
                    const std::string &ext,
                    std::vector<::btool::node::Node *> *objs) {
  DEBUGS() << "collect from src " << n->name() << std::endl;

  auto obj_name = ::btool::util::string::Replace(n->name(), ext, ".o");
  auto obj = s->Get(obj_name);
  if (obj == nullptr) {
    DEBUGS() << "cannot find obj for obj name " << obj_name << std::endl;
    assert(0);
  }
  if (::btool::util::Contains(*objs, obj)) {
    return;
  }
  objs->push_back(obj);

  for (auto d : *n->dependencies()) {
    DEBUGS() << "considering dependency " << d->name() << std::endl;
    if (::btool::util::string::HasSuffix(d->name().c_str(), ".h")) {
      auto src_name = ::btool::util::string::Replace(d->name(), ".h", ext);
      auto src = s->Get(src_name);
      if (src == nullptr) {
        DEBUGS() << "no src for src_name " << src_name << std::endl;
      } else {
        CollectObjects(s, src, ext, objs);
      }
    }
  }
}

void CollectLibraries(::btool::app::collector::Store *s, ::btool::node::Node *n,
                      std::vector<::btool::node::Node *> *libs) {
  std::vector<std::string> lib_names;
  ::btool::app::collector::CollectStringsProperties(
      n, &lib_names,
      [](const ::btool::node::PropertyStore *ps)
          -> const std::vector<std::string> * {
        return Properties::Libraries(ps);
      });

  for (const auto &lib_name : lib_names) {
    auto lib = s->Get(lib_name);
    if (lib == nullptr) {
      DEBUGS() << "unknown lib for lib_name " << lib_name << std::endl;
      assert(0);
    }

    libs->push_back(lib);
  }
}

void CollectLinkFlags(::btool::node::Node *n, std::vector<std::string> *flags) {
  ::btool::app::collector::CollectStringsProperties(
      n, flags,
      [](const ::btool::node::PropertyStore *ps)
          -> const std::vector<std::string> * {
        return Properties::LinkFlags(ps);
      });
}

}  // namespace btool::app::collector::cc
